Separate xsd documentation and appinfo contents on hover and completion

xorye opened the pull request:
Fixes #630 

This PR is accompanied by https://github.com/redhat-developer/vscode-xml/pull/260

Tests must be written.

Demo:

xml:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root
attributeNameOnlyDocumentation='onlyDocumentation'
attributeNameOnlyAppinfo='onlyAppinfo'
attributeNameBoth='bothDocumentationAndAppinfo'
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:noNamespaceSchemaLocation="test.xsd">
<elementOnlyDocumentation></elementOnlyDocumentation>
<elementOnlyAppinfo></elementOnlyAppinfo>
<elementBoth></elementBoth>
</root>
```

xsd:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<aa:schema xmlns:aa="http://www.w3.org/2001/XMLSchema">

  <aa:simpleType name="attributeValue">
    <aa:restriction base="aa:string">
      <aa:enumeration value="onlyDocumentation">
        <aa:annotation>
          <aa:documentation>attribute value documentation</aa:documentation>
        </aa:annotation>
      </aa:enumeration>
      <aa:enumeration value="onlyAppinfo">
        <aa:annotation>
          <aa:documentation>attribute value appinfo</aa:documentation>
        </aa:annotation>
      </aa:enumeration>
      <aa:enumeration value="bothDocumentationAndAppinfo">
        <aa:annotation>
          <aa:documentation>attribute value documentation</aa:documentation>
          <aa:appinfo>attribute value appinfo</aa:appinfo>
        </aa:annotation>
      </aa:enumeration>
    </aa:restriction>
  </aa:simpleType>

<aa:element name="root">
  <aa:complexType>
    <aa:sequence>
      <aa:element name="elementOnlyDocumentation">
        <aa:annotation>
          <aa:documentation>element only documentation element only documentation element only documentation element only documentation</aa:documentation>
        </aa:annotation>
      </aa:element>
      <aa:element name="elementOnlyAppinfo">
        <aa:annotation>
          <aa:appinfo>element only app info element only app info element only app info element only app info element only app info</aa:appinfo>
        </aa:annotation>
      </aa:element>
      <aa:element name="elementBoth">
        <aa:annotation>
          <aa:documentation>element only documentation element only documentation element only documentation element only documentation</aa:documentation>
          <aa:appinfo>element only app info element only app info element only app info element only app info element only app info</aa:appinfo>
        </aa:annotation>
      </aa:element>
    </aa:sequence>
   
    <aa:attribute name="attributeNameOnlyDocumentation" type="attributeValue" use="required">
      <aa:annotation>
        <aa:documentation>attribute name documentation</aa:documentation>
      </aa:annotation>
    </aa:attribute>
    <aa:attribute name="attributeNameOnlyAppinfo" type="attributeValue" use="required">
      <aa:annotation>
        <aa:appinfo>attribute name appinfo</aa:appinfo>
      </aa:annotation>
    </aa:attribute>
    <aa:attribute name="attributeNameBoth" type="attributeValue" use="required">
      <aa:annotation>
        <aa:documentation>attribute name documentation</aa:documentation>
        <aa:appinfo>attribute name appinfo</aa:appinfo>
      </aa:annotation>
    </aa:attribute>
  </aa:complexType>
</aa:element>

</aa:schema>
```
![demo gif](https://raw.githubusercontent.com/xorye/gifs/master/pr/hover_completion.gif?token=AE3CR5IYZEGXIQQE6TN5GR26Y2I74)


Signed-off-by: David Kwon <dakwon@redhat.com>

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/CMAttributeDeclaration.java:
```diff
@@ -31,9 +33,9 @@
 	Collection<String> getEnumerationValues(); 
 	
-	String getDocumentation();
+	String getDocumentation(ISharedSettingsRequest settings);
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/CMAttributeDeclaration.java:
```diff
@@ -31,9 +33,9 @@
+	String getDocumentation(ISharedSettingsRequest settings);
 
-	String getValueDocumentation(String value);
+	String getValueDocumentation(String value, ISharedSettingsRequest settings);
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XSDDocumentationSource.java:
```diff
@@ -0,0 +1,20 @@
+
+import com.google.gson.annotations.SerializedName;
+
+public enum XSDDocumentationSource {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLPreferences.java:
```diff
@@ -66,7 +71,24 @@ public QuoteStyle getQuoteStyle() {
+	 * 
+	 * @param docSource
+	 */
+	public void setDocumentationSource(XSDDocumentationSource docSource) {
```
rename to documentationSource

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+import org.eclipse.lemminx.settings.XSDDocumentationSource;
+import org.eclipse.lemminx.utils.StringUtils;
+
+public class XSDDocumentation {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+		this.prefix = XSDAnnotationModel.getPrefix(annotations, value);
+	}
+
+	public String getFormattedDocumentation(boolean html) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+		return result.toString();
+	}
+
+	public String getDocumentation() {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+		return documentation;
+	}
+
+	public void setDocumentation(String documentation) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+		this.documentation = documentation;
+	}
+
+	public String getAppinfo() {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+		return appinfo;
+	}
+
+	public void setAppinfo(String appinfo) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	private String applyPrefix(String documentation, boolean html) {
+		StringBuilder result = new StringBuilder();
```
I wonder if xs:documentation, xs:appinfo should be embed inside a p element to have a cariage return after the xs documentation. @fbricon what do you think about that?

xorye replied:
In my latest push, there is a new line after xs:documentation and xs:appinfo

angelozerr replied:
great thanks.

xorye commented:
My latest commits add more tests (markdown + plaintext) and adds the ability to show all `documentation` & `appinfo` content in cases where there is more than one inside of a single `annotation` element: 

```
<xs:annotation>
          <xs:documentation>first element documentation</xs:documentation>
          <xs:documentation>second element documentation</xs:documentation>
          <xs:documentation>third element documentation</xs:documentation>
          <xs:appinfo>first element appinfo</xs:appinfo>
          <xs:appinfo>second element appinfo</xs:appinfo>
          <xs:appinfo>third element appinfo</xs:appinfo>
</xs:annotation>
```

Demo:
![image](https://user-images.githubusercontent.com/20326645/82579534-f3a18700-9b5b-11ea-97db-842e44e6de3b.png)

Here is a gif:
![demo](https://raw.githubusercontent.com/xorye/gifs/master/pr/hover_doc_appinfo.gif?token=AE3CR5IZ6DUREY54C2A5GKC6Z7VJG)


Here is the updated xml and xsd content used in the demo:
[test.xml](https://gist.github.com/xorye/4ed54c448fb032e8c65a92c960cb213f)
[test.xsd](https://gist.github.com/xorye/3f2903e66f3b0795e9d779521dd41f4d)

xorye commented:
> adds the ability to show all documentation & appinfo content in cases where there is more than one inside of a single annotation element

To enhance this feature it would be great if the documentation would show the `source` attribute, if it exists. According to this: https://www.w3schools.com/xml/el_documentation.asp the `documentation` could optionally have `source` and `xml:lang` attributes. (If there is a better website for XML docs, please let me know). The `source` attribute is optional for `appinfo` as well: https://www.w3schools.com/xml/el_appinfo.asp.

I mention this because in pom.xml, lots of elements have two `documentation` elements.

For example in maven-4.0.0.xsd:
```
<xs:element minOccurs="0" name="groupId" type="xs:string">
  <xs:annotation>
    <xs:documentation source="version">3.0.0+</xs:documentation>
    <xs:documentation source="description">
      
      A universally unique identifier for a project. It is normal to
      use a fully-qualified package name to distinguish it from other
      projects with a similar name (eg. &lt;code&gt;org.apache.maven&lt;/code&gt;).
      
    </xs:documentation>
  </xs:annotation>
</xs:element>
```

The documentation provided in this PR will visually distinguish content from `<xs:documentation source="version">` and `<xs:documentation source="version">`:

![image](https://user-images.githubusercontent.com/20326645/82591809-e55c6680-9b6d-11ea-9cfb-cbb16c112620.png)

Could this be done in a different PR?

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:
```diff
@@ -291,7 +289,10 @@ public void onAttributeName(boolean generateValue, ICompletionRequest request, I
 			CMElementDeclaration cmElement, boolean canSupportSnippet, boolean generateValue,
-			ICompletionRequest request, ICompletionResponse response, SharedSettings sharedSettings) {
+			ICompletionRequest request, ICompletionResponse response) {
+		if (cmElement == null) {
```
No need to do this check. It is done before calling this method..

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java:
```diff
@@ -60,32 +67,47 @@ public String getDefaultValue() {
 		XSObjectList annotations = getValueAnnotations();
-		documentation = XSDAnnotationModel.getDocumentation(annotations, value);
+		boolean markdownSupported = request.canSupportMarkupKind(MarkupKind.MARKDOWN);
+		documentation = (new XSDDocumentation(annotations, value, docStrategy, !markdownSupported))
```
Remove extra `(` (the first.

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java:
```diff
@@ -35,12 +38,16 @@
+
+	private String tagDocumentation;
 
 	private Map<String, String> valuesDocumentation;
```
final

xorye replied:
Currently, I cannot set to final because it gets reassigned here: 

https://github.com/eclipse/lemminx/blob/1080bf5860753537aea27a612e486dc35b0041f1/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java#L106-L109

angelozerr replied:
replace with 

`this.valuesDocumentation .clear();`

angelozerr commented on org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/XMLSchemaHoverDocumentationAppinfoTest.java:
```diff
@@ -0,0 +1,330 @@
+/**
```
Change header with red hat + update the year to 2020

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java:
```diff
@@ -274,29 +274,30 @@ private static String getFileName(String schemaURI) {
 	 * @param cmAttribute
 	 * @param ownerElement
-	 * @param support
+	 * @param request
```
Add Javadoc for each parameters

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java:
```diff
@@ -35,12 +38,16 @@
 	private final XSAttributeUse attributeUse;
-	private String documentation;
+
+	private String tagDocumentation;
```
rename to attrNameDocumentation

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
 class XSDAnnotationModel {
 
-	String appInfo;
+	List<String> appInfo;
```
private final

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
+	List<String> appInfo;
 
-	String documentation;
+	List<String> documentation;
```
private final

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
 	}
 
-	public String getAppInfo() {
+	public List<String> getAppInfo() {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
 	}
 
-	public String getDocumentation() {
+	public List<String> getDocumentation() {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
 	}
 
-	public static String getDocumentation(XSObjectList annotations) {
+	public static List<String> getDocumentation(XSObjectList annotations) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
 	}
 
-	public static String getDocumentation(XSObjectList annotations, String value) {
+	public static List<String> getDocumentation(XSObjectList annotations, String value) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
+		return result;
+	}
+
+	public static List<String> getAppInfo(XSObjectList annotations) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
+		return getAppInfo(annotations, null);
+	}
+
+	public static List<String> getAppInfo(XSObjectList annotations, String value) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -38,35 +42,75 @@
+		return appinfo;
+	}
+
+	private static XSAnnotation getXSAnnotation(Object annotation, String value) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -76,32 +120,35 @@ public static String getDocumentation(XSObjectList annotations, String value) {
-						doc.append(getDocumentation(annotationString));
-					}
-				}
+	public static String getPrefix(XSObjectList annotations) {
```
Add Javadoc (by given a sample)

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -76,32 +120,35 @@ public static String getDocumentation(XSObjectList annotations, String value) {
+		return getPrefix(annotations, null);
+	}
+
+	public static String getPrefix(XSObjectList annotations, String value) {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDAnnotationModel.java:
```diff
@@ -159,6 +206,12 @@ public void characters(char[] ch, int start, int length) throws SAXException {
 			super.characters(ch, start, length);
 		}
 
+		private void addIfNonEmptyString(List<String> list, String str) {
```
private static

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,139 @@
+	 */
+	public String getFormattedDocumentation(boolean html) {
+		StringBuilder result = new StringBuilder();
+		result.append(getFormatted("documentation", documentation, html));
```
Create a constant (private static final) for "documentation" and "appinfo".

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,139 @@
+		return result.toString().trim();
+	}
+
+	private void convertToPlainText(List<String> list) {
```
private static

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -0,0 +1,139 @@
+		}
+	}
+
+	private String getFormatted(String prefix, List<String> content, boolean html) {
```
private static

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/ISharedSettingsRequest.java:
```diff
@@ -0,0 +1,29 @@
+
+import org.eclipse.lemminx.settings.SharedSettings;
+
+public interface ISharedSettingsRequest {
```
Add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XSDDocumentationSource.java:
```diff
@@ -0,0 +1,23 @@
+ * XSD documentation source enum
+ */
+public enum XSDDocumentationSource {
+	@SerializedName("Both") both,
```
Why are you using SerializedName?

xorye replied:
I guess I don't necessarily need it for `both`, but I use it for `documentation` and `appinfo`:
https://github.com/xorye/lsp4xml/blob/914dc9e2c530e7d9c269dfebc3934de849ad0719/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XSDDocumentationSource.java#L21-L22

because the enum value in vscode-xml is like so:
https://github.com/redhat-developer/vscode-xml/blob/e5be58f2c56dbb7132297086534811ce0cbff3f9/package.json#L238-L239

Maybe instead, we should remove `@SerializedName` and change the enum names to `schemaDocumentation` and `schemaAppinfo`?

angelozerr replied:
I find XSDDocumentationSource is not very understandable and I think enum should short and without spaces. Here a suggestion:

```json
"xml.preferences.showShemaDocumentationType": {
          "type": "string",
          "default": "documentation",
          "enum": [
            "documentation",
            "appinfo",
            "both"
          ]
...

@fbricon what do you think about that?

fbricon replied:
documentation, appinfo, all

angelozerr replied:
I agree that all is better, but what do you think about 

 * renaming `XSDDocumentationSource `to `ShemaDocumentationType`
 * renaming `xml.preferences.documentationSource` to `xml.preferences.showShemaDocumentationType`

fbricon replied:
ok.

also need a `none` choice

xorye replied:
fixed

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XSDDocumentationSource.java:
```diff
@@ -0,0 +1,23 @@
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * XSD documentation source enum
```
Please add in the Javadoc that this enumeration manages the show of documentation, appinfo or both

angelozerr commented:
> Could this be done in a different PR?

Yes please create an issue for that.

xorye commented:
Thank you for the amazing feedback.

I updated the settings in the client-side PR https://github.com/redhat-developer/vscode-xml/pull/260:
![settings gif](https://raw.githubusercontent.com/xorye/gifs/master/pr/hover_doc_appinfo_setting.gif?token=AE3CR5NIKF6JTONRJQKPOVK62UORS)

angelozerr commented:
There is null for appinfo:

![image](https://user-images.githubusercontent.com/1932211/82903838-da387a80-9f61-11ea-962b-b08c25cacc88.png)

xorye commented:
Updated default value to `all` and updated when the titles show up:
![demo](https://raw.githubusercontent.com/xorye/gifs/master/pr/hover_updated.gif?token=AE3CR5LVFFZ6J6UWBCT3RHS62Z7K2)

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLPreferences.java:
```diff
@@ -66,7 +71,30 @@ public QuoteStyle getQuoteStyle() {
+	/**
+	 * Returns the XSD documentation source
+	 */
+	public SchemaDocumentationType getDocumentationSource() {
```
replace with getShowSchemaDocumentationType()

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLPreferences.java:
```diff
@@ -66,7 +71,30 @@ public QuoteStyle getQuoteStyle() {
+	 * 
+	 * @param documentationSource
+	 */
+	public void setDocumentationSource(SchemaDocumentationType documentationSource) {
```
replace with replace with setShowSchemaDocumentationType
